app/src/main/java/org/thoughtcrime/securesms/jobs/FastJobStorage.java

```diff
@@ -0,0 +1,396 @@
+        iter.set(updated);
+      }
+    }
+  }
+
+  @Override
+  public synchronized void updateJobAfterRetry(@NonNull String id, boolean isRunning, int runAttempt, long nextRunAttemptTime, @NonNull String serializedData) {
+    JobSpec job = getJobById(id);
+    if (job == null || !job.isMemoryOnly()) {
+      jobDatabase.updateJobAfterRetry(id, isRunning, runAttempt, nextRunAttemptTime, serializedData);
+    }
+
+    ListIterator<JobSpec> iter = jobs.listIterator();
+
+    while (iter.hasNext()) {
+      JobSpec existing = iter.next();
+      if (existing.getId().equals(id)) {
+        JobSpec updated = new JobSpec(existing.getId(),
+                                      existing.getFactoryKey(),
+                                      existing.getQueueKey(),
+                                      existing.getCreateTime(),
+                                      nextRunAttemptTime,
+                                      runAttempt,
+                                      existing.getMaxAttempts(),
+                                      existing.getLifespan(),
+                                      serializedData,
+                                      existing.getSerializedInputData(),
+                                      isRunning,
+                                      existing.isMemoryOnly());
+        iter.set(updated);
+      }
+    }
+  }
+
+  @Override
+  public synchronized void updateAllJobsToBePending() {
+    jobDatabase.updateAllJobsToBePending();
+
+    ListIterator<JobSpec> iter = jobs.listIterator();
+
+    while (iter.hasNext()) {
+      JobSpec existing = iter.next();
+      JobSpec updated  = new JobSpec(existing.getId(),
+                                     existing.getFactoryKey(),
+                                     existing.getQueueKey(),
+                                     existing.getCreateTime(),
+                                     existing.getNextRunAttemptTime(),
+                                     existing.getRunAttempt(),
+                                     existing.getMaxAttempts(),
+                                     existing.getLifespan(),
+                                     existing.getSerializedData(),
+                                     existing.getSerializedInputData(),
+                                     false,
+                                     existing.isMemoryOnly());
+      iter.set(updated);
+    }
+  }
+
+  @Override
+  public void updateJobs(@NonNull List<JobSpec> jobSpecs) {
+    List<JobSpec> durable = new ArrayList<>(jobSpecs.size());
+    for (JobSpec update : jobSpecs) {
+      JobSpec found = getJobById(update.getId());
+      if (found == null || !found.isMemoryOnly()) {
+        durable.add(update);
+      }
+    }
+
+    if (durable.size() > 0) {
+      jobDatabase.updateJobs(durable);
+    }
+
+    Map<String, JobSpec>  updates = Stream.of(jobSpecs).collect(Collectors.toMap(JobSpec::getId));
+    ListIterator<JobSpec> iter    = jobs.listIterator();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **281**. At this line the following field is not protected by synchronization: `jobs`. Your code typically protects this field with synchronization on object: `FastJobStorage.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `FastJobStorage.this`. For example, like at lines: **86**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **55**, **78**, **86**, **191**, **220**, **246**, and **314**.